sys/sched.c

```diff
@@ -59,8 +59,11 @@ void sched_switch(thread_t *newtd) {
 
   if (newtd == NULL) {
     newtd = runq_choose(&runq);
-    if (newtd)
+    if (newtd) {
+      if(newtd->td_type == TD_USER && newtd->user_map)
+          set_active_vm_map(newtd->user_map);
```

Comment:
> Shouldn't we introduce process structure? This shouldn't be that hard. 

I don't think it's necessary to introduce process structure. How our kernel would benefit from it right now? After all, each thread may own its private user address space. Note that L4 microkernels don't have a concept of process at all – only threads and address spaces.

> Besides I don't see how to change vm_map without calling set_active_vm_map in ctx_switch, which would require change from LEAF to NESTED, is this a problem, or you have better idea?

Calling `set_active_vm_map` (which is our version of `pmap_activate`) seems to be fine. Can't tell if change to `NESTED` is a problem in our case – that's something to be verified.

Let's return to my previous post for a while. For sure, after doing a context switch between threads working in kernel-space, the chosen thread has to be able to modify user-space memory (e.g. copy data to user space in order to return from `read` syscall).